core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixTextToSpeech.kt

```diff
@@ -0,0 +1,295 @@
+import org.kiwix.kiwixmobile.core.utils.LanguageUtils.Companion.iSO3ToLocale
+import org.kiwix.kiwixmobile.core.utils.TAG_KIWIX
+import java.util.HashMap
+import java.util.concurrent.atomic.AtomicInteger
+
+/**
+ * Constructor.
+ *
+ * @param context the context to create TextToSpeech with
+ * @param onInitSucceedListener listener that receives event when initialization of TTS is done
+ * (and does not receive if it failed)
+ * @param onSpeakingListener listener that receives an event when speaking just started or
+ */
+class KiwixTextToSpeech internal constructor(
+  val context: Context,
+  onInitSucceedListener: OnInitSucceedListener,
+  val onSpeakingListener: OnSpeakingListener,
+  private val onAudioFocusChangeListener: OnAudioFocusChangeListener,
+  private val zimReaderContainer: ZimReaderContainer
+) {
+  private val focusLock: Any = Any()
+  private val am: AudioManager = context.getSystemService(Context.AUDIO_SERVICE) as AudioManager
+  @JvmField var currentTTSTask: TTSTask? = null
+  private lateinit var tts: TextToSpeech
+
+  /**
+   * Returns whether the TTS is initialized.
+   *
+   * @return `true` if TTS is initialized; `false` otherwise
+   */
+  private var isInitialized = false
+
+  init {
+    Log.d(TAG_KIWIX, "Initializing TextToSpeech")
+    initTTS(onInitSucceedListener)
+  }
+
+  private fun initTTS(onInitSucceedListener: OnInitSucceedListener) {
+    tts = TextToSpeech(instance, OnInitListener { status: Int ->
+      if (status == TextToSpeech.SUCCESS) {
+        Log.d(TAG_KIWIX, "TextToSpeech was initialized successfully.")
+        this.isInitialized = true
+        onInitSucceedListener.onInitSucceed()
+      } else {
+        Log.e(TAG_KIWIX, "Initialization of TextToSpeech Failed!")
+        context.toast(R.string.texttospeech_initialization_failed, Toast.LENGTH_SHORT)
+      }
+    })
+  }
+
+  /**
+   * Reads the currently selected text in the WebView.
+   */
+  fun readSelection(webView: WebView) {
+    webView.loadUrl("javascript:tts.speakAloud(window.getSelection().toString());", null)
+  }
+
+  /**
+   * Starts speaking the WebView content aloud (or stops it if TTS is speaking now).
+   */
+  fun readAloud(webView: WebView) {
+    val isPaused = currentTTSTask?.paused
+    if (isPaused == true) {
+      onSpeakingListener.onSpeakingEnded()
+      currentTTSTask = null
+    } else if (tts.isSpeaking) {
+      if (tts.stop() == TextToSpeech.SUCCESS) {
+        tts.setOnUtteranceProgressListener(null)
+        onSpeakingListener.onSpeakingEnded()
+      }
+    } else {
+      val locale = iSO3ToLocale(zimReaderContainer.language)
+      var result = 0
+      if ("mul" == zimReaderContainer.language) {
+        Log.d(
+          TAG_KIWIX, "TextToSpeech: disabled " +
+            zimReaderContainer.language
+        )
+        context.toast(R.string.tts_not_enabled, Toast.LENGTH_LONG)
+        return
+      }
+      if (locale == null || tts.isLanguageAvailable(locale)
+          .also {
```

Comment:
this is quite unpleasant to read, an if statement  shouldn't be preforming assignment in the middle. I guess the original was doing this though.

how about

```
if (locale==null || isMissingOrUnsupported(tts.isLanguageAvailable(locale)) )
```

Reply:
@macgills Is this ok?
```
  if (locale == null || isMissingOrUnsupportedLanguage(tts.isLanguageAvailable(locale))) {

Reply:
all good but I would use something more descriptive than `data: Int`, `languageAvailabilityResult`/`textToSpeechAvailability`? Bit long but more accurate probably if we statically import the constants it all maybe fits on 1 line still. Also brackets are unnecessary in `isMissingOrUnsupportedLanguage`

Reply:
> if we statically import the constants it all maybe fits on 1 line still

@macgills I am sorry. I don't understand how to make constant from this. Can you help me with how to do it? and in which file i save this constant?

Reply:
`TextToSpeech.LANG_MISSING_DATA` this is one of the constants I am talking about, place your cursor either before or after the `.` and then press `alt+enter` you should get an option to statically import